Fix configure when --with-cpu not supplied

pmur opened the pull request:
This will now attempt to use the newest CPU architecture available
from the compiler, and prevent a degenerate case whereby hardware
dfp is supported, but a specific PPC machine is not selected.

pmur commented:
This fixes issue #19

tuliom commented:
This patch is interesting, but I believe we should have some code first to force soft dfp when with-cpu isn't specified.
Only after that we could have a patch that tries to automatically get the cpu value like this one.

tuliom commented on configure.ac:
```diff
@@ -220,13 +220,37 @@ AC_ARG_WITH([cpu],
-      ]])], [with_dfp=yes], [with_dfp=no]) ;;
+      ]])], [with_dfp=yes; submachine=power6], [with_dfp=no])
+    AC_COMPILE_IFELSE([AC_LANG_SOURCE([[
+      #ifdef _ARCH_PWR6X
```
Shouldn't you test for _ARCH_PWR6 ?

tuliom commented on configure.ac:
```diff
@@ -220,13 +220,37 @@ AC_ARG_WITH([cpu],
+      #endif
+      ]])], [with_dfp=yes; submachine=power6x], [])
+    AC_COMPILE_IFELSE([AC_LANG_SOURCE([[
+      #ifdef _ARCH_PWR7
```
Do we really need to test for POWER7 and POWER8?
They both set _ARCH_PWR6 and _ARCH_PWR7.

pmur replied:
Yes, the order here matters. We test for increasingly newer processors until we fail.

Suggestions for a more canonical way to test for these values? Ideally, we might just run one test which would print out the correct values based on the supplied macros.

tuliom commented:
> Yes, the order here matters. We test for increasingly newer processors until we fail.

I wasn't really referring to the order, but if it was really required to test for POWER7 and POWER8. But now I see I was still blindly following the old behavior that didn't use to change `submachine`.
Let me put it in a different way: to find the value of `with_dfp`, you only need to test for POWER6.

> Suggestions for a more canonical way to test for these values? Ideally, we might just run one test which would print out the correct values based on the supplied macros.

You could encapsulate the duplicated code, so that you could have something like this:
```
CMACRO_IFELSE([_ARCH_PWR6],
              [with_dfp=yes; submachine=power6],
              [with_dfp=no]);
CMACRO_IFELSE([_ARCH_PWR6X],
              [submachine=power6x], );
CMACRO_IFELSE([_ARCH_PWR7],
              [submachine=power7], );
...
```

pmur commented:
I've updated the patch to simplify the macro tests as suggested, and added a new patch to pass -mno-hard-dfp when building for soft_dfp on power/z targets.

tuliom commented:
Both patches LGTM in overall.  You may merge it yourself after:
 - Adding a signed-off-by tag to all of them (according to https://github.com/libdfp/libdfp/blob/master/CONTRIBUTING#L50-L52)
 - Improving the output of ```configure --help``` to explain the usage of ```--with-cpu=no```.

pmur commented:
Merged with suggested changes.
